File: leetcode/c++/101_symmetric_tree.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (root == nullptr) return true;
        return isSymmetric(root->left, root->right);
    }

    bool isSymmetric(TreeNode* l, TreeNode* r) {
        if (!l || !r) return l == r;
        return l->val == r->val 
            && isSymmetric(l->left, r->right)
            && isSymmetric(l->right, r->left);
    }
};

File: leetcode/python/101_symmetric_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSymmetric(self, root: TreeNode | None) -> bool:
        if not root:
            return True
        return self.helper(root.left, root.right)
    
    def helper(self, l: TreeNode | None, r: TreeNode | None) -> bool:
        if not l and not r:
            return True
        elif not l or not r or l.val != r.val:
            return False
        return self.helper(l.left, r.right) and self.helper(l.right, r.left)